src/storage/rdb_listpack.cc

```diff
@@ -137,88 +120,66 @@ uint32_t ListPack::encodeBackLen(uint32_t len) {
-    GET_OR_RET(peekOK(5));
-    value_len = (static_cast<uint32_t>(data[pos_])) | (static_cast<uint32_t>(data[pos_ + 1]) << 8) |
-                (static_cast<uint32_t>(data[pos_ + 2]) << 16) | (static_cast<uint32_t>(data[pos_ + 3]) << 24);
-    // skip 5byte encoding type
-    pos_ += 5;
-    GET_OR_RET(peekOK(value_len));
-    value = input_.substr(pos_, value_len);
-    // skip the value bytes and the length of the element
-    pos_ += value_len + encodeBackLen(value_len + 5);
-  } else if (c == ListPackEOF) {
-    ++pos_;
+  uint32_t entry_bytes = 0;
+
+  auto encoding = stream_.Read<uint8_t>();
+  if ((encoding & ListPack7BitUIntMask) == ListPack7BitUInt) {
+    // 0|xxxxxxx - 7bit unsigned integer
+    num = encoding & 0x7F;
+    value = std::to_string(num);
+    entry_bytes = 1;
+  } else if ((encoding & ListPack6BitStringMask) == ListPack6BitString) {
+    // 10xxxxxx - 6bit string
+    len = (encoding & 0x3F);
+    value = stream_.Read(len);
+    entry_bytes = 1 + len;
+  } else if ((encoding & ListPack13BitIntMask) == ListPack13BitInt) {
+    // 110|xxxxx yyyyyyyy -- 13 bit signed integer
+    uint16_t u16 = (encoding & 0x1F) << 8 | stream_.Read<uint8_t>();
+    num = static_cast<int16_t>(u16 << 3) >> 3;
+    value = std::to_string(num);
+    entry_bytes = 1 + 1;
+  } else if ((encoding & ListPack16BitIntMask) == ListPack16BitInt) {
+    // 1111|0001 <16 bits signed integer>
+    num = stream_.Read<int16_t>();
+    value = std::to_string(num);
+    entry_bytes = 1 + sizeof(int16_t);
+  } else if ((encoding & ListPack24BitIntMask) == ListPack24BitInt) {
+    // 1111|0010 <24 bits signed integer>
+    uint32_t u32 = stream_.Read<uint8_t>() << 8 | stream_.Read<uint8_t>() << 16 | stream_.Read<uint8_t>() << 24;
+    num = static_cast<int32_t>(u32) >> 8;
+    value = std::to_string(num);
+    entry_bytes = 1 + 3;
+  } else if ((encoding & ListPack32BitIntMask) == ListPack32BitInt) {
+    // 1111|0011 <32 bits signed integer>
+    num = stream_.Read<int32_t>();
+    value = std::to_string(num);
+    entry_bytes = 1 + sizeof(int32_t);
+  } else if ((encoding & ListPack64BitIntMask) == ListPack64BitInt) {
+    // 1111|0100 <64 bits signed integer>
+    num = stream_.Read<int64_t>();
+    value = std::to_string(num);
+    entry_bytes = 1 + sizeof(int64_t);
+  } else if ((encoding & ListPack12BitStringMask) == ListPack12BitString) {
+    // 1110|xxxx yyyyyyyy -- string with length up to 4095
+    len = (encoding & 0xF) << 8 | stream_.Read<uint8_t>();
+    value = stream_.Read(len);
+    entry_bytes = 2 + len;
+  } else if ((encoding & ListPack32BitStringMask) == ListPack32BitString) {
+    // 1111|0000 <4 bytes len> <large string>
+    len = stream_.Read<uint32_t>();
+    value = stream_.Read(len);
+    entry_bytes += 1 + sizeof(uint32_t) + len;
   } else {
-    return {Status::NotOK, "invalid listpack entry"};
+    throw std::runtime_error("invalid listpack entry");
```

Comment:
I don't think it is reasonable to change from `StatusOr` back to exceptions here and only here, since it break the consistency of error handling.